### PR TITLE
Reproduce lecture materials

### PR DESCRIPTION
**Pull request by khaletska**

### CODE REVIEW

File: Card.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+//  Created by Yeliena Khaletska on 03.06.2024.
+//
+
+import Foundation
+
+struct Card
+{
+    var isFaceUp = false
+    var isMatched = false
+    var identifier: Int
```
Comment (Shiaulis) [inline Card.swift]:
```suggestion
    let identifier: Int
```

File: Concentration/Concentration.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+//
+
+import Foundation
+
+class Concentration
+{
+    var cards = [Card]()
+    private var indexOfOneFaceUpCard: Int?
+
+    func chooseCard(at index: Int) {
```
Comment (Shiaulis) [inline Concentration/Concentration.swift:24]: let's have functions under init

File: Concentration/Concentration.swift [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        if !cards[index].isMatched {
```
Comment (Shiaulis) [inline Concentration/Concentration.swift]:
```suggestion
        if !self.cards[index].isMatched {
```

File: Concentration/Concentration.swift [hunk 4]
```diff
@@ -0,0 +1,45 @@
+//
+//  Concentration.swift
+//  Concentration
+//
+//  Created by Yeliena Khaletska on 03.06.2024.
+//
+
+import Foundation
+
+class Concentration
```
Comment (Shiaulis) [inline Concentration/Concentration.swift]:
```suggestion
final class Concentration
```

File: Concentration/Concentration.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        if !cards[index].isMatched {
+            if let matchIndex = indexOfOneFaceUpCard, matchIndex != index {
+                // check if cards match
+                if cards[matchIndex].identifier == cards[index].identifier {
+                    cards[matchIndex].isMatched = true
+                    cards[index].isMatched = true
+                }
+                cards[index].isFaceUp = true
+                indexOfOneFaceUpCard = nil
+            } else {
```
Comment (Shiaulis) [inline Concentration/Concentration.swift]:
```suggestion
            } 
            else {
```

File: Concentration/ViewController.swift [hunk 6]
```diff
@@ -7,13 +7,52 @@
...
+        self.flipCount += 1
+        if let cardIndex = self.cardButtons.firstIndex(of: sender) {
+            game.chooseCard(at: cardIndex)
+            updateUI()
+        }
+    }
+
+    private func updateUI() {
+        for index in cardButtons.indices {
+            let button = cardButtons[index]
```
Comment (Shiaulis) [inline Concentration/ViewController.swift]:
```suggestion
            let button = self.cardButtons[index]
```

File: Concentration/ViewController.swift [hunk 7]
```diff
@@ -7,13 +7,52 @@
...
+            } else {
+                button.setTitle("", for: UIControl.State.normal)
+                button.backgroundColor = card.isMatched ? UIColor.clear : UIColor.systemOrange
+            }
+        }
+    }
+
+    private func emoji(for card: Card) -> String {
+        if emoji[card.identifier] == nil, emojiChoices.count > 0 {
+            let randIndex = Int(arc4random_uniform(UInt32(emojiChoices.count)))
```
Comment (Shiaulis) [inline Concentration/ViewController.swift]: let's improve this logic with modern `Int.random()`

**Review by Shiaulis — CHANGES_REQUESTED** [review]
let's address minor issues and merge this PR 🙏🏾

**Comment by khaletska** [comment]
I'll add more improvements based on our discussion of @Shiaulis PR in the new branch along with the homework

**Review by Shiaulis — APPROVED** [review]
tested, works good 👌🏾 🙏🏾

**Pull request merged**